### ISSUE TITLE
strncpy error

### ISSUE THREAD

**Issue by loveshack** [jamesodhunt/procenv#7]
Jeff Law, of Red Hat, sent me this concerning the Fedora package, and his analysis looks right, though it looks as if the buffer length will actually always be OK.

I can fix the build in Fedora as necessary, but thought you should get first dibs on a patch.

```
procenv is going to fail to build in Fedora when the compiler team drops
in the upcoming gcc-9 rpms.

In particular gcc-9 flags this mistake:

> BUILDSTDERR: In file included from /usr/include/string.h:494,
> BUILDSTDERR:                  from ./procenv.h:33,
> BUILDSTDERR:                  from procenv.c:29:
> BUILDSTDERR: In function 'strncpy',
> BUILDSTDERR:     inlined from 'format_time' at procenv.c:4340:2:
> BUILDSTDERR: /usr/include/bits/string_fortified.h:106:10: error:
'__builtin_strncpy' specified bound depends on the length of the source
argument [-Werror=stringop-overflow=]
> BUILDSTDERR:   106 |   return __builtin___strncpy_chk (__dest, __src,
__len, __bos (__dest));
> BUILDSTDERR:       |
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> BUILDSTDERR: procenv.c: In function 'format_time':
> BUILDSTDERR: procenv.c:4334:6: note: length computed here
> BUILDSTDERR:  4334 |  l = strlen (str);
> BUILDSTDERR:       |      ^~~~~~~~~~~~
> BUILDSTDERR: cc1: all warnings being treated as errors


Essentially it's telling you the code is nonsense.

       l = strlen (str);

        if (len < l)
                bug ("buffer too small");

        /* Ensure nul byte copied */
        strncpy (buffer, str, l+1);

From my reading I really wonder if the check should have been

if (len < l + 1)
  bug ("buffer too small");

strcpy (buffer, str);


The code is clearly trying to DTRT, it just gets it wrong and as a
result is potentially buffer overflows AFAICT.
```

**Comment by xnox** [comment]
This now also affects Ubuntu.

**Comment by vorlonofportland** [comment]
Opened https://github.com/jamesodhunt/procenv/pull/9 which I believe is a correct fix for this (fixing a real off-by-one bug in addition to the gcc linting error).

**Comment by jamesodhunt** [comment]
Thanks to @vorlonofportland this was fixed in 54dd0b726d46f1caf26623ce952a0e7bfb1f9671.
